Save failure details to result file

mike-nguyen opened the pull request:
Added a callback_plugin that makes the logs more human readable and
also saves down failure details to a file.

A majority of this code is actually from other projects.  The code
that alters the Ansible output is from openshift-ansible which is
derived from an upstream project (in the comments).  I had to override
the baseclass methods to track the failure details.  In order to
preserve the default Ansible logging, I copied the original method
code back into the override method.

All test that want to leverage the callback_plugin need a symlink to
the top callback_plugins directory otherwise the default Ansible logging
will be used.

miabbott commented on callback_plugins/default.py:
```diff
@@ -0,0 +1,145 @@
+
+        # Save result file in current working directory
+        try:
+            os.remove('result')
```
I wonder if we want to make the filename more unique.  In our own Jenkins jobs, we should get a fresh workspace for each time we run the tests, but if other folks run this test multiple times from the same directory, we'll end up overwriting the file.

Perhaps prepend it with the playbook name and/or append a timestamp?

miabbott replied:
Maybe even go the extra mile and let the filename be configurable via environment variable?

miabbott commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) [CI] Report the role which failed in the email notifications for the [a-h-t] CI jobs](https://trello.com/c/SMOBZjPj/401-3-ci-report-the-role-which-failed-in-the-email-notifications-for-the-a-h-t-ci-jobs)

jlebon commented:
This is pretty cool. Just looking at the PR tester output, you can tell it's a drastic difference. Once suggestion might be to prefix `STDOUT` and `STDERR` lines with e.g. `> ` or maybe just a `---` header and footer to make it more easy to tell what the actual output is.

mike-nguyen commented:
Added an environment variable (AHT_RESULT_FILE) that, if set, will create a result file that will be blank if it passes or contain the failure task/role/output if it fails.  Also added some headers and footers to the stdout/stderr/msg.

miabbott commented:
The sample output from the Red Hat CI check looks great with this plugin.  I'm OK merging this assuming @jlebon has no other comments?

jlebon commented:
No complaints here! 👍

mike-nguyen commented:
Thanks!
